text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,156 @@
+
+Implementing a new columnar format is complex and costly (multiple language implementations, tests, optimizations, industry adoption). Reusing Apache Arrow is an interesting approach to mitigate this issue.
+
+## Prior art and alternatives
+
+* [Column-oriented DBMS](https://en.wikipedia.org/wiki/Column-oriented_DBMS)
+* [Apache Arrow](https://arrow.apache.org/)
+
+## Open questions
+
+More work needs to be done around examplars and histograms representation.
+
+More discussions should happen on the processor and storage layers. This approach could simplify significantly the design of OTEL compatible stream processing and database systems.
+
+## Future possibilities
+
+* Leverage Apache Arrow dictionary
+
+## Appendices
+
+### event proto
+
+Protobuf specification (draft) for an Arrow-based OpenTelemetry event.
+
+```protobuf
+syntax = "proto3";
+
+package opentelemetry.proto.arrow_events.v1;
+
+import "opentelemetry/proto/common/v1/common.proto";
+import "opentelemetry/proto/resource/v1/resource.proto";
+
+option java_multiple_files = true;
+option java_package = "io.opentelemetry.proto.events.v1";
+option java_outer_classname = "EventsProto";
+option go_package = "github.com/open-telemetry/opentelemetry-proto/gen/go/events/v1";
+
+// A column-oriented collection of events from a Resource.
+message ResourceEvents {
+  // The resource for the events in this message.
+  // If this field is not set then no resource info is known.
+  opentelemetry.proto.resource.v1.Resource resource = 1;
+
+  // A list of events that originate from a resource.
+  repeated InstrumentationLibraryEvents instrumentation_library_events = 2;
+
+  // This schema_url applies to the data in the "resource" field. It does not apply
+  // to the data in the "instrumentation_library_events" field which have their own
+  // schema_url field.
+  string schema_url = 3;
+}
+
+// A collection of Events produced by an InstrumentationLibrary.
+message InstrumentationLibraryEvents {
+  // The instrumentation library information for the events in this message.
+  // Semantically when InstrumentationLibrary isn't set, it is equivalent with
+  // an empty instrumentation library name (unknown).
+  opentelemetry.proto.common.v1.InstrumentationLibrary instrumentation_library = 1;
+
+  // A list of batch of events that originate from an instrumentation library.
+  repeated BatchEvent batches = 2;
+
+  // dropped_events_count is the number of dropped events. If the value is 0, then no
+  // events were dropped.
+  uint32 dropped_events_count = 3;
+}
+
+// A typed collection of events with a columnar-oriented representation.
+// All the events in this collection share the same schema url.
+message BatchEvent {
+  // This schema_url applies to all events in this batch.
+  string schema_url = 1;
+
+  uint32 size = 2;
+
+  bytes  arrow_buffer = 5;
```

Comment:
This is an opaque byte array. How are Otel metrics, spans and logs represented here? I am not familiar with Apache Arrow but it appears to be a generic columnar data representation format, not a telemetry data format. So, how exactly are metrics, spand and logs represented as Apache Arrow byte sequence?

Reply:
Thanks for the review. I will add a section describing the proposed mapping between metrics, logs and traces to the corresponding Arrow schemas. Part of it is already in the code for the benchmark. I need to make it more visible and complete.

Reply:
Added a mapping section for each OTEL entities + a link to the Arrow memory layout.
Regarding the benchmark, I collected data points for different batch sizes (i.e. 10, 100, 500, 1000, 5000, 10000).

Reply:
Although I signed the CLA, I don't understand why Github continues to think that this is not the case.

Reply:
Added my answers directly in the "OpenTelemetry entities to Arrow mapping" section